Userland/Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -148,15 +149,164 @@ void Document::removed_last_ref()
 }
 
 // https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#dom-document-write
-void Document::write(Vector<String> const& strings)
+ExceptionOr<void> Document::write(Vector<String> const& strings)
 {
-    dbgln("TODO: document.write({})", strings);
+    // 1. If document is an XML document, then throw an "InvalidStateError" DOMException.
+    if (doctype() && doctype()->name() == "xml") {
+        return DOM::InvalidStateError::create("Write() called on XML document.");
+    }
+
+    // 2. If document's throw-on-dynamic-markup-insertion counter is greater than 0, then throw an "InvalidStateError" DOMException.
+    if (m_throw_on_dynamic_markup_insertion_counter > 0) {
+        return DOM::InvalidStateError::create("throw-on-dynamic-markup-insertion-counter greater then zero.");
+    }
+
+    // 3. If document's active parser was aborted is true, then return.
+    if (m_active_parser_was_aborted) {
+        return {};
+    }
+
+    // 4. If the insertion point is undefined, then:
+    if (!(m_parser && m_parser->tokenizer()->is_insertion_point_defined())) {
+        //      1. If document's unload counter is greater than 0 or document's ignore-destructive-writes counter is greater than 0, then return.
+        if (m_unload_counter > 0 || m_ignore_destructive_writes_counter > 0) {
+            return {};
+        }
+        //      2. Run the document open st with document.
```

Comment:
Typo: st -> steps

Reply:
Additionally, please remove all that whitespace: `// 2. Run ...`